addon/@ember/test-waiters/wait-for.ts

```diff
@@ -96,3 +135,33 @@ function isThenable(
     typeof maybePromise.then === 'function'
   );
 }
+
+function isGenerator(
+  maybeGenerator: any | CoroutineGenerator<unknown>
+): maybeGenerator is CoroutineGenerator<unknown> {
+  return typeof maybeGenerator[Symbol.iterator] === 'function';
+}
+
+const GENERATOR_WAITER = buildWaiter('@ember/test-waiters:generator-waiter');
+
+function waitForGenerator<T>(
+  generator: CoroutineGenerator<T>,
+  label?: string
+): CoroutineGenerator<T> {
+  GENERATOR_WAITER.beginAsync(generator, label);
+
+  return {
+    next() {
+      try {
+        let val = generator.next();
+        if (val.done) {
+          GENERATOR_WAITER.endAsync(generator);
+        }
+        return val;
+      } catch (e) {
+        GENERATOR_WAITER.endAsync(generator);
```

Comment:
Lets use `finally` instead of `catch` here (no need to re-throw, and the devtools "stop on uncaught exceptions" will still work)

Reply:
See big comment below -- pretty sure we need this catch block since this try...catch happens once per iteration/`yield` statement, not once per full-iteration-over-the-generator-function, i.e. (in `ember-concurrency` terms) once per yielded promise, not once per task execution.

Reply:
This makes sense, @bendemboski. I'm a bit concerned about calling `endAsync` here mid-iteration, as it may open the testing system up to unpause waiting and incorrectly signal `settled`. WDYT?

Reply:
I think we're in the uncanny valley between thinking of this in terms of general iterator semantics and thinking of it in terms of a coroutine/ember-concurrency task.

I suppose in terms of general iterator semantics, the user might catch this exception but not discard the iterator, then fix some state or something, and then continue iteration.

But in the context of coroutines and `ember-concurrency`, this means that the coroutine/task has terminated by throwing an exception, and execution will not resume. This is the equivalent of an exception propagating out of an async function, causing the promise it returned to reject, and irreversibly terminating the execution of the function. So this is really the equivalent of [this](https://github.com/emberjs/ember-test-waiters/blob/cccdb301f3/addon/%40ember/test-waiters/wait-for-promise.ts#L44), and the fact that the user could theoretically resume iteration is just a side-effect of the fact that we're using iterator semantics to implement async behavior.

Reply:
So, we are making the assumption here that the wrapping/management code interprets a thrown exception as "terminate execution," but in all of the contexts that we can think of this being used, that's accurate, and I don't think we can possibly do without that assumption (because then we have no idea if/when the user will resume iteration, so we might never have another opportunity `endAsync`, causing `settled()` to hang forever)